Use of deprecated fluid_synth_set_sample_rate

## Version of OpenTTD
master

## Expected result
Compiles without warnings

## Actual result
```
/home/lordaro/dev/openttd/src/music/fluidsynth.cpp:85:53: warning: ‘void fluid_synth_set_sample_rate(fluid_synth_t*, float)’ is deprecated [-Wdeprecated-declarations]
   85 |  fluid_synth_set_sample_rate(_midi.synth, samplerate);
      |                                                     ^
In file included from /usr/include/fluidsynth.h:97,
                 from /home/lordaro/dev/openttd/src/music/fluidsynth.cpp:16:
/usr/include/fluidsynth/synth.h:179:38: note: declared here
  179 | FLUID_DEPRECATED FLUIDSYNTH_API void fluid_synth_set_sample_rate(fluid_synth_t *synth, float sample_rate);
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

## Steps to reproduce
Compile

From http://www.fluidsynth.org/api/synth_8h.html#ac89771633b393c807e5d5766f5a219b0 :
>As of fluidsynth 2.1.0 this function has been deprecated. Changing the sample-rate is generally not considered to be a real-time use-case, as it always produces some audible artifact ("click", "pop") on the dry sound and effects (because LFOs for chorus and reverb need to be reinitialized). The sample-rate change may also require memory allocation deep down in the effect units. However, this memory allocation may fail and there is no way for the caller to know that, because the actual change of the sample-rate is executed during rendering. This function cannot (must not) do the sample-rate change itself, otherwise the synth needs to be locked down, causing rendering to block. Esp. do not use this function if this synth instance is used by an audio driver, because the audio driver cannot be notified by this sample-rate change. Long story short: don't use it.

--- a/src/music/fluidsynth.cpp
+++ b/src/music/fluidsynth.cpp
@@ -15,11 +15,13 @@
 #include "midifile.hpp"
 #include <fluidsynth.h>
 #include "../mixer.h"
+#include <mutex>
 
 static struct {
 	fluid_settings_t* settings;    ///< FluidSynth settings handle
 	fluid_synth_t* synth;          ///< FluidSynth synthesizer handle
 	fluid_player_t* player;        ///< FluidSynth MIDI player handle
+	std::mutex synth_mutex;        ///< Guard mutex for synth access
 } _midi; ///< Metadata about the midi we're playing.
 
 /** Factory for the FluidSynth driver. */
@@ -42,12 +44,16 @@ static const char *default_sf[] = {
 
 static void RenderMusicStream(int16 *buffer, size_t samples)
 {
-	if (!_midi.synth || !_midi.player) return;
+	std::unique_lock<std::mutex> lock{ _midi.synth_mutex, std::try_to_lock };
+
+	if (!lock.owns_lock() || !_midi.synth || !_midi.player) return;
 	fluid_synth_write_s16(_midi.synth, samples, buffer, 0, 2, buffer, 1, 2);
 }
 
 const char *MusicDriver_FluidSynth::Start(const char * const *param)
 {
+	std::lock_guard<std::mutex> lock{ _midi.synth_mutex };
+
 	const char *sfont_name = GetDriverParam(param, "soundfont");
 	int sfont_id;
 
@@ -58,6 +64,11 @@ const char *MusicDriver_FluidSynth::Start(const char * const *param)
 	if (!_midi.settings) return "Could not create midi settings";
 	/* Don't try to lock sample data in memory, OTTD usually does not run with privileges allowing that */
 	fluid_settings_setint(_midi.settings, "synth.lock-memory", 0);
+
+	/* Install the music render routine and set up the samplerate */
+	uint32 samplerate = MxSetMusicSource(RenderMusicStream);
+	fluid_settings_setnum(_midi.settings, "synth.sample-rate", samplerate);
+	DEBUG(driver, 1, "Fluidsynth: samplerate %.0f", (float)samplerate);
 
 	/* Create the synthesizer. */
 	_midi.synth = new_fluid_synth(_midi.settings);
@@ -81,19 +92,23 @@ const char *MusicDriver_FluidSynth::Start(const char * const *param)
 
 	_midi.player = nullptr;
 
-	uint32 samplerate = MxSetMusicSource(RenderMusicStream);
-	fluid_synth_set_sample_rate(_midi.synth, samplerate);
-	DEBUG(driver, 1, "Fluidsynth: samplerate %.0f", (float)samplerate);
-
 	return nullptr;
 }
 
 void MusicDriver_FluidSynth::Stop()
 {
 	MxSetMusicSource(nullptr);
-	this->StopSong();
-	delete_fluid_synth(_midi.synth);
-	delete_fluid_settings(_midi.settings);
+
+	std::lock_guard<std::mutex> lock{ _midi.synth_mutex };
+
+	if (_midi.player != nullptr) delete_fluid_player(_midi.player);
+	_midi.player = nullptr;
+
+	if (_midi.synth != nullptr) delete_fluid_synth(_midi.synth);
+	_midi.synth = nullptr;
+
+	if (_midi.settings != nullptr) delete_fluid_settings(_midi.settings);
+	_midi.settings = nullptr;
 }
 
 void MusicDriver_FluidSynth::PlaySong(const MusicSongInfo &song)
@@ -105,6 +120,8 @@ void MusicDriver_FluidSynth::PlaySong(const MusicSongInfo &song)
 	if (filename.empty()) {
 		return;
 	}
+
+	std::lock_guard<std::mutex> lock{ _midi.synth_mutex };
 
 	_midi.player = new_fluid_player(_midi.synth);
 	if (!_midi.player) {
@@ -128,6 +145,8 @@ void MusicDriver_FluidSynth::PlaySong(const MusicSongInfo &song)
 
 void MusicDriver_FluidSynth::StopSong()
 {
+	std::lock_guard<std::mutex> lock{ _midi.synth_mutex };
+
 	if (!_midi.player) return;
 
 	fluid_player_stop(_midi.player);
@@ -142,6 +161,7 @@ void MusicDriver_FluidSynth::StopSong()
 
 bool MusicDriver_FluidSynth::IsSongPlaying()
 {
+	std::lock_guard<std::mutex> lock{ _midi.synth_mutex };
 	if (!_midi.player) return false;
 
 	return fluid_player_get_status(_midi.player) == FLUID_PLAYER_PLAYING;
@@ -149,6 +169,9 @@ bool MusicDriver_FluidSynth::IsSongPlaying()
 
 void MusicDriver_FluidSynth::SetVolume(byte vol)
 {
+	std::lock_guard<std::mutex> lock{ _midi.synth_mutex };
+	if (_midi.settings == nullptr) return;
+
 	/* Allowed range of synth.gain is 0.0 to 10.0 */
 	/* fluidsynth's default gain is 0.2, so use this as "full
 	 * volume". Set gain using OpenTTD's volume, as a number between 0

